doc/man/luerl.3
.\" Automatically generated by Pandoc 3.6
.\"
.TH "luerl" "3" "2018\-2024" ""
.SH Name
luerl \- The basic interface to the Luerl system
.SH Interface functions \- New Version
The \f[B]Lua State\f[R] parameter is the state of a Lua VM instance.
It must be created with the \f[B]luerl:init()\f[R] call and be carried
from one call to the next.
.PP
As it is possible in Lua to create self\-referencing data structures,
indeed the standard libraries have many instances of this, then using
the functions which decode their return values will generate an error
when they would cause an infinite loop during the decoding.
An simple example is the top level table which contains a key
\f[B]\f[CB]_G\f[B]\f[R] which references the top\-level table.
.PP
Note that Lua \f[B]Chunks\f[R] (see definition below) can travel between
different States.
They are precompiled bits of code, independent of State.
That you can `carry around' this is no unique to Luerl but a low\-level
implementation detail of the standard Lua \c
.UR https://lua.org
language
.UE \c
, for more on chunks \c
.UR https://www.lua.org/manual/5.3/manual.html#3.3.2
read
.UE \c
\ the official Lua 5.3 \c
.UR https://www.lua.org/manual/5.3/manual.html
reference manual
.UE \c
\&.
.SS Spec Definitions
\f[B]Binary\f[R] means an Erlang binary string.
.PD 0
.P
.PD
\f[B]Chunks\f[R] means a portion of precompiled bytecode.
.PD 0
.P
.PD
\f[B]State\f[R] means a Lua State, this \f[I]is\f[R] a Lua VM instance.
.PD 0
.P
.PD
\f[B]Path\f[R] means a file system path and file name.
.PD 0
.P
.PD
\f[B]KeyPath\f[R] means an Erlang list of \f[B]atoms\f[R] representing
nested names, e.g.\ [table,pack] for table.pack.
.PD 0
.P
.PD
\f[B]Keys\f[R] means Lua table keys, the keys of a key\-value structure.
.PP
\f[B]CompileOptions\f[R] means a list of compiler options.
Currently supported options are `return', which returns the errors and
warnings, and `report' which will log the errors and warnings.
.PP
\f[B]LuaCallReturn = {ok, Result, State} | {lua_error, Error,
State}\f[R]
.PD 0
.P
.PD
This is the return value from evaluating a Lua call.
.SS Functions
.SS luerl:init() \-> State
Get a new Lua State = a fresh Lua VM instance.
.SS luerl:gc(State) \-> State
Runs the garbage collector on a state and returns the new state.
.SS luerl:load(String|Binary[, CompileOptions], State) \-> {ok, Function, State} | CompileError
Parse a Lua chunk as string or binary, and return a compiled chunk
(`form').
.SS luerl:loadfile(FileName[, CompileOptions], State) \-> {ok, Function, State} | CompileError
Parse a Lua file, and return a compiled chunk (`form').
.SS luerl:path_loadfile([Path, ], FileName[, CompileOptions], State) \-> {ok,Function,FullName,State} | {error, Reason}
Search Path until the file FileName is found.
Parse the file and return a compiled chunk (`form').
If Path is not given then the path defined in the environment variable
LUA_LOAD_PATH is used.
.SS luerl:load_module(KeyPath, ErlangModule, State) \-> State
Load \f[CR]ErlangModule\f[R] and install its table at \f[CR]KeyPath\f[R]
which is \f[B]NOT\f[R] encoded.
.SS luerl:load_module_dec(EncodedKeyPath, ErlangModule, State) \-> State
Load \f[CR]ErlangModule\f[R] and install its table at \f[CR]KeyPath\f[R]
which is encoded.
.SS luerl:do(String|Binary|Form, State) \-> {ok, Result, NewState} | {lua_error, Error, State} | CompileError
Evaluate a Lua expression and return its result which is \f[B]NOT\f[R]
decoded, and the new Lua State.
.SS luerl:do_dec(String|Binary|Form, State) \-> {ok, Result, NewState} | {lua_error, Error, State} | CompileError
Evaluate a Lua expression and return its result which is automatically
decoded, and the new Lua State.
.SS luerl:dofile(Path, State) \-> {ok, Result, NewState} | {lua_error, Error, State} | CompileError
Load and execute the Lua code in the file and return its result which is
\f[B]NOT\f[R] decoded, and the new Lua State.
Equivalent to doing luerl:do(\[lq]return dofile(`FileName')\[rq]).
.SS luerl:dofile_dec(Path[, State]) \-> {ok, Result, NewState} | {lua_error, Error, State} | CompileError
Load and execute the Lua code in the file and return its result which is
automatically decoded, and the new Lua State.
.SS luerl:call(FuncRef, ArgRefs, State) \-> {ok, Result, State}
.SS luerl:call_chunk(FuncRef, ArgRefs, State) \-> {ok, Result, State} | {lua_error, Error, State}
Call a compiled chunk or function.
Use the call_chunk, call has been kept for backwards compatibility.
.SS luerl:call_function(FuncRef | FuncPath, ArgRefs, State] \-> {ok, Result, State} | {lua_error, Error, State}
Call a function already defined in the state.
\f[CR]Result\f[R] is \f[B]NOT\f[R] decoded.
.SS luerl:call_function_dec(KeyPath, Args, State) \-> {ok, Result, State} | {lua_error, Error, State}
Call a function already defined in the state.
\f[CR]KeyPath\f[R] is a list of keys to the function.
\f[CR]KeyPath\f[R], \f[CR]Args\f[R] and \f[CR]Result\f[R] are
automatically encoded/decoded.
.SS luerl:call_method(ObjRef, Method, ArgRefs, State) \-> {ok, Result, State} | {lua_error, Error, State}
Call a method already defined in the state.
.SS luerl:call_method_dec(KeyPath, Method, Args, State) \-> {ok, Result, State} | {lua_error, Error, State}
Call a method already defined in the state.
\f[CR]KeyPath\f[R] is a list of keys to the method.
\f[CR]KeyPath\f[R], \f[CR]Method\f[R], \f[CR]Args\f[R] and
\f[CR]Result\f[R] are automatically encoded/decoded.
.SS luerl:get_table_keys(KeyPath, State) \-> {ok, Result, State} | {lua_error, Error, State}
Gets a value inside the Lua state.
\f[CR]KeyPath\f[R] and \f[CR]Result\f[R] are \f[B]NOT\f[R]
encoded/decoded.
.SS luerl:get_table_keys_dec(KeyPath, State) \-> {ok, Result, State} | {lua_error, Error, State}
Gets a value inside the Lua state.
\f[CR]KeyPath\f[R] is automatically encoded and \f[CR]Result\f[R] is
decoded.
.SS luerl:set_table_keys(KeyPath, Value, State) \-> {ok,State} | {lua_error, Error, State}
Sets a value inside the Lua state.
\f[CR]KeyPath\f[R] and \f[CR]Value\f[R] are \f[B]NOT\f[R] encoded.
.SS luerl:set_table_keys_dec(KeyPath, Value, State) \-> {ok, Result, State} | {lua_error, Error, State}
Sets a value inside the Lua state.
\f[CR]KeyPath\f[R] and \f[CR]Value\f[R] are automatically encoded and
\f[CR]Result\f[R] is decoded.
.SS luerl:get_table_key(Table, Key, State) \-> {ok, Result, State} | {lua_error, Error, State}
Gets the value of a key in a table.
\f[CR]Table\f[R] and \f[CR]Key\f[R] are \f[B]NOT\f[R] encoded and
\f[CR]Result\f[R] is \f[B]NOT\f[R] decoded.
.SS luerl:set_table_key(Table, Key, Value, State) \-> {ok, State} | {lua_error, Error, State}
Sets the value of a key in a table.
\f[CR]Table\f[R], \f[CR]Key\f[R] and \f[CR]Value\f[R] are \f[B]NOT\f[R]
encoded.
.SS luerl:get_stacktrace(State) \-> [{FuncName,{file,FileName},{line,Line}}]
Return a stack trace of the current call stack in the state.
.SS luerl:encode(Term, State) \-> {LuerlTerm,State}
Encode the Erlang representation of a term into Luerl form updating the
state when necessary.
.SS luerl:encode_list([Term], State) \-> {[LuerlTerm],State}
Encode a list of Erlang term representations into a list of Luerl forms
updating the state when necessary.
.SS luerl:decode(LuerlTerm, State) \-> Term
Decode a term in the Luerl form into its Erlang representation.
.SS luerl:decode_list([LuerlTerm], State) \-> [Term]
Decode a list of Luerl terms into a list of Erlang representations.
.SS luerl:put_private(Key, Term, State) \-> State.
Puts a private value under key that is not exposed to the runtime.
.SS luerl:get_private(Key, State) \-> Term.
Get a private value for the given key.
.SS luerl:delete_private(Key, State) \-> Term.
Deletes the private value for the given key.
.SS Passing String in the Erlang Shell
Here we are going to look at passing in command strings into the
\f[CR]luerl:do/2\f[R] function, especially when these strings contain
Lua strings.
The problem is to make sure that the strings in the Lua commands are
processes correctly.
.PP
First just just doing it from the standard Lua shell as the source of
truth:
.IP
.EX
>return \[dq]aéb\[rs]235c\[dq]
aéb?c
>string.byte(\[dq]aéb\[rs]235c\[dq], 1, 20)
97    195    169    98    235    99
.EE
.PP
Now doing it from Erlang where I need to use \[rs]\[rs] to get a \[rs]
into the string:
.IP
.EX
1> St = luerl:init(), ok.
ok
2> f(S), S = \[dq]return \[aq]aéb\[rs]\[rs]235c\[aq]\[dq].
[114,101,116,117,114,110,32,39,97,233,98,92,50,51,53,99,39]
3> f(R), {ok,R,_} = luerl:do(S, St), io:write(R).
[<<97,195,169,98,235,99>>]
.EE
.PP
Now doing it from Elixir using the \f[CR]\[ti]C\f[R] sigil to get the
handling of the string right:
.IP
.EX
iex(1)> st = Luerl.init(); :ok
:ok
iex(4)> s = \[ti]C\[dq]return \[aq]aéb\[rs]235c\[aq]\[dq]
[114,101,116,117,114,110,32,39,97,233,98,92,50,51,53,99,39]
iex(3)> {ok,r,_} = Luerl.do(st, s); r
[<<97, 195, 169, 98, 235, 99>>]
.EE
.PP
So it works as expected.
We use \f[CR]io:write\f[R] to write out the characters in the binary
without any parsing or trying to be smart.
.SH AUTHORS
Jean Chassoul, Robert Virding.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Changed

- `luerl:get_private` returns `{ok, Val} | error` tuple

### Fixed

- files with only comments can now be loaded
- atoms are now decoded as strings


[unreleased]: https://github.com/olivierlacan/keep-a-changelog/compare/v1.3.0...HEAD
[1.3.0]: https://github.com/rvirding/luerl/compare/v1.2.3...v1.3.0
[1.2.3]: https://github.com/rvirding/luerl/compare/v1.2.2...v1.2.3


src/luerl.erl
%% Copyright (c) 2020-2024 Robert Virding
%%
%% Licensed under the Apache License, Version 2.0 (the "License");
%% you may not use this file except in compliance with the License.
%% You may obtain a copy of the License at
%%
%%     http://www.apache.org/licenses/LICENSE-2.0
%%
%% Unless required by applicable law or agreed to in writing, software
%% distributed under the License is distributed on an "AS IS" BASIS,
%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
%% See the License for the specific language governing permissions and
%% limitations under the License.

%% File    : luerl_new.erl
%% Authors : Robert Virding
%% Purpose : The new basic LUA 5.3 interface.

-module(luerl).

-include("luerl.hrl").

?MODULEDOC("""
Luerl is an implementation of Lua 5.3 written in Erlang.
This is the main public API module for interfacing with Luerl.

For Elixir users, the `Elixir.Luerl` module provides an idiomatic interface
with state as the first argument for better pipe operator usage.
""").

%% Basic user API to luerl.
-export([init/0,gc/1,
         load/2,load/3,loadfile/2,loadfile/3,
         path_loadfile/2,path_loadfile/3,path_loadfile/4,
         load_module/3,load_module_dec/3,
         do/2,do_dec/2,do/3,do_dec/3,
         dofile/2,dofile/3,dofile_dec/2,dofile_dec/3,
         call/3,call_chunk/2,call_chunk/3,
         call_function/3,call_function_enc/3,call_function_dec/3,
         call_method/4,call_method_dec/4,
         get_table_keys/2,get_table_keys_dec/2,
         set_table_keys/3,set_table_keys_dec/3,
         get_table_key/3,set_table_key/4,
         get_stacktrace/1
        ]).

%% Tracing.
-export([set_trace_func/2,clear_trace_func/1,
         set_trace_data/2,get_trace_data/1]).

%% Encoding and decoding.
-export([encode/2,encode_list/2,decode/2,decode_list/2]).

%% Helping with storing VM state
-export([externalize/1,internalize/1]).

%% Storing and retrieving private data
-export([put_private/3,get_private/2,delete_private/2]).

%% init() -> State.

init() ->
    luerl_emul:init().

%% gc(State) -> State.
gc(St) ->
    luerl_heap:gc(St).

%% set_trace_func(TraceFunction, State) -> State.
%% clear_trace_func(State) -> State.
%% get_trace_data(State) -> TraceData.
%% set_trace_data(TraceData, State) -> State.
%%  Set the trace function and access the trace data.

set_trace_func(Tfunc, St) ->
    St#luerl{trace_func=Tfunc}.

clear_trace_func(St) ->
    St#luerl{trace_func=none}.

get_trace_data(St) ->
    St#luerl.trace_data.

set_trace_data(Tdata, St) ->
    St#luerl{trace_data=Tdata}.

%% load(String|Binary, State) -> {ok,FuncRef,NewState}.
%% load(String|Binary, Options, State) -> {ok,FuncRef,NewState}.

load(Bin, St) -> load(Bin, [return], St).

load(Bin, Opts, St) when is_binary(Bin) ->
    load(binary_to_list(Bin), Opts, St);
load(Str, Opts, St0) ->
    case luerl_comp:string(Str, Opts) of
        {ok,Chunk} ->
            {FunRef,St1} = luerl_emul:load_chunk(Chunk, St0),
            {ok,FunRef,St1};
        Error ->                                %Compile error
            Error
    end.

%% loadfile(FileName, State) -> {ok,FuncRef,NewState}.
%% loadfile(FileName, Options, State) -> {ok,FuncRef,NewState}.

loadfile(Name, St) -> loadfile(Name, [return], St).

loadfile(Name, Opts, St0) ->
    case luerl_comp:file(Name, Opts) of
        {ok,Chunk} ->
            {Func,St1} = luerl_emul:load_chunk(Chunk, St0),
            {ok,Func,St1};
        Error -> Error
    end.

%% path_loadfile(FileName, State) -> {ok,Function,FullName,State}.
%% path_loadfile(Path, FileName, State) -> {ok,Function,FullName,State}.
%% path_loadfile(Path, FileName, Options, State) ->
%%     {ok,Function,FullName,State}.
%%  When no path is given we use the value of LUA_LOAD_PATH.
%%  We manually step down the path to get the correct handling of
%%  filenames by the compiler.

path_loadfile(Name, St) ->
    Path = case os:getenv("LUA_LOAD_PATH") of
               false -> [];                     %You get what you asked for
               Env ->
                   %% Get path separator depending on os type.
                   Sep = case os:type() of
                             {win32,_} -> ";";  %Windows
                             _ -> ":"           %Unix
                         end,
                   string:tokens(Env, Sep)      %Split into path list
           end,
    path_loadfile(Path, Name, [return], St).

path_loadfile(Dirs, Name, St) ->
    path_loadfile(Dirs, Name, [return], St).

path_loadfile([Dir|Dirs], Name, Opts, St0) ->
    Full = filename:join(Dir, Name),
    case loadfile(Full, Opts, St0) of
        {ok,Func,St1} ->
            {ok,Func,Full,St1};
        {error,[{_,_,enoent}],_} ->             %Couldn't find the file
            path_loadfile(Dirs, Name, St0);
        Error -> Error
    end;
path_loadfile([], _, _, _) ->
    {error,[{none,file,enoent}],[]}.

%% load_module(LuaTablePath, ModuleName, State) -> State.
%%  Load module and add module table to the path.

load_module([_|_] = Lfp, Mod, St0) ->
    {Tab,St1} = Mod:install(St0),
    luerl_emul:set_table_keys(Lfp, Tab, St1);
load_module(_, _, _) ->
    error(badarg).

%% load_module_dec(DecodedTablePath, ModuleName, State) -> State.
%%  Load module and add module table to the path.

load_module_dec([_|_] = Dfp, Mod, St0) ->
    {Efp,St1} = encode_list(Dfp, St0),
    load_module(Efp, Mod, St1);
load_module_dec(_, _, _) ->
    error(badarg).

%% luerl:do(String|Binary|Form, State) ->
%% luerl:do(String|Binary|Form, CompileOptions, State) ->
%%     {ok,Result,NewState} | {lua_error,Error,State}.

do(S, St) -> do(S, [return], St).

do(S, Opts, St0) ->
    case load(S, Opts, St0) of
        {ok,Func,St1} ->
            call_function(Func, [], St1);
        Error -> Error
    end.

do_dec(S, St) ->
    do_dec(S, [return], St).

do_dec(S, Opts, St0) ->
    case do(S, Opts, St0) of
        {ok,Eret,St1} ->
            {ok,decode_list(Eret, St1),St1};
        Error -> Error
    end.

%% luerl:dofile(FileName, State) ->
%% luerl:dofile(FileName, CompileOptions, State) ->
%%     {ok,Result,NewState} | {lua_error,Error,State}.

dofile(File, St) -> dofile(File, [], St).

dofile(File, Opts, St0) ->
    case loadfile(File, Opts, St0) of
        {ok,Func,St1} ->
            call_function(Func, [], St1);
        Error -> Error
    end.

dofile_dec(File, St) ->
    dofile_dec(File, [], St).

dofile_dec(File, Opts, St0) ->
    case dofile(File, Opts, St0) of
        {ok,Eret,St1} ->
            {ok,decode_list(Eret, St1),St1};
        Error -> Error
    end.

%% call(FuncRef, Args, State) ->
%% call_chunk(FuncRef, State) ->
%% call_chunk(FuncRef, Args, State) ->
%%     {ok,Return,State} | {lua_error,Error,State}.

call(C, As, St) ->
    call_function(C, As, St).

call_chunk(C, St) ->
    call_chunk(C, [], St).

call_chunk(C, As, St) ->
    call_function(C, As, St).

%% call_function(LuaFuncRef | LuaTablePath, Args, State) ->
%%     {ok,LuaReturn,State} | {lua_error,Error,State}.

call_function(Epath, Args, St0) when is_list(Epath) ->
    {ok,Efunc,St1} = get_table_keys(Epath, St0),
    call_function(Efunc, Args, St1);
call_function(Func, Args, St0) ->
    try
        {Ret,St1} = luerl_emul:functioncall(Func, Args, St0),
        {ok,Ret,St1}
    catch
        error:{lua_error,_E,_St} = LuaErr ->
            LuaErr
    end.

%% call_function_enc(DecodedFuncRef, Args, State) ->
%%     {ok,LuaReturn,State} | {lua_error,Error,State}.

call_function_enc(Dtpath, Dargs, St0) ->
    {Epath,St1} = encode_list(Dtpath, St0),
    {Eargs,St2} = encode_list(Dargs, St1),
    call_function(Epath, Eargs, St2).

%% call_function_dec(DecodedFuncRef, Args, State) ->
%%     {ok,DecodedReturn,State} | {lua_error,Error,State}.

call_function_dec(Dtpath, Dargs, St0) ->
    case call_function_enc(Dtpath, Dargs, St0) of
        {ok,Eret,St1} ->
            {ok,decode_list(Eret, St1),St1};
        LuaError -> LuaError
    end.

%% call_method(LuaObject, Method, Args, State) ->
%%     {ok,Return,State} | {lua_error,Error,State}.

call_method(Obj, Meth, Args, St0) ->
    try
        {Ret,St1} = luerl_emul:methodcall(Obj, Meth, Args, St0),
        {ok,Ret,St1}
    catch
        error:{lua_error,_E,_St} = LuaErr ->
            LuaErr
    end.

%% call_method_dec(DecodedObject, Method, Args, State) ->
%%     {ok,DecodedReturn,State} | {lua_error,Error,State}.

call_method_dec(Dobj, Dmeth, Dargs, St0) ->
    {ok,Eobj,St1} = get_table_keys_dec(Dobj, St0),
    {Emeth,St2} = encode(Dmeth, St1),
    {Eargs,St3} = encode_list(Dargs, St2),
    case call_method(Eobj, Emeth, Eargs, St3) of
        {ok,Eret,St4} ->
            {ok,decode_list(Eret, St4),St4};
        LuaError -> LuaError
    end.

%% get_table_keys(Keys, State) ->
%% get_table_keys_dec(DecodedKeys, State) ->
%%     {ok,Return,State} | {lua_error,Error,State}.
%% set_table_keys(Keys, Val, State) ->
%% set_table_keys_dec(DecodedKeys, DecodedVal, State) ->
%%     {ok,State} | {lua_error,Error,State}.

get_table_keys(Keys, St0) ->
    try
        {Eret,St1} = luerl_emul:get_table_keys(Keys, St0),
        {ok,Eret,St1}
    catch
        error:{lua_error,_E,_St} = LuaErr ->
            LuaErr
    end.

get_table_keys_dec(Dkeys, St0) ->
    {Ekeys,St1} = encode_list(Dkeys, St0),
    case get_table_keys(Ekeys, St1) of
        {ok,Eret,St2} ->
            {ok,decode(Eret, St2),St2};
        LuaError -> LuaError
    end.

set_table_keys(Keys, Val, St0) ->
    try
        St1 = luerl_emul:set_table_keys(Keys, Val, St0),
        {ok,St1}
    catch
        error:{lua_error,_E,_St} = LuaErr ->
            LuaErr
    end.

set_table_keys_dec(Dkeys, Dval, St0) ->
    {Ekeys,St1} = encode_list(Dkeys, St0),
    {Eval,St2} = encode(Dval, St1),
    set_table_keys(Ekeys, Eval, St2).

%% get_table_key(Tab, Key, State) ->
%%     {ok,Value,State} | {lua_error,Error,State}.
%% set_table_key(Tab, Key, Value, State) ->
%%     {ok,State} | {lua_error,Error,State}.

get_table_key(Tab, Key, St0) ->
    try
        {Eret,St1} = luerl_emul:get_table_key(Tab, Key, St0),
        {ok,Eret,St1}
    catch
        error:{lua_error,_E,_St} = LuaErr ->
            LuaErr
    end.

set_table_key(Tab, Key, Val, St0) ->
    try
        St1 = luerl_emul:set_table_key(Tab, Key, Val, St0),
        {ok,St1}
    catch
        error:{lua_error,_E,_St} = LuaErr ->
            LuaErr
    end.

%% get_stacktrace(State) -> [{FuncName,[{file,FileName},{line,Line}]}].

get_stacktrace(#luerl{cs=Stack}=St) ->
    Fun = fun (Frame, Acc) -> do_stackframe(Frame, Acc, St) end,
    {_,Trace} = lists:foldl(Fun, {1,[]}, Stack),
    lists:reverse(Trace).

do_stackframe(#call_frame{func=Funref,args=Args}, {Line,Trace}, St) ->
    case Funref of
        #funref{} ->
            {Func,_} = luerl_heap:get_funcdef(Funref, St),
            Anno = Func#lua_func.anno,
            Name = case luerl_anno:get(name, Anno) of
                       undefined -> <<"-no-name-">>;
                       N -> N
                   end,
            File = luerl_anno:get(file, Anno),
            {Line,[{Name,Args,[{file,File},{line,Line}]} | Trace]};
        #erl_func{code=Fun} ->
            {module,Module} = erlang:fun_info(Fun, module),
            {name,Name} = erlang:fun_info(Fun, name),
            FileName = get_filename(Module),
            {Line,[{{Module,Name},Args,[{file,FileName}]} | Trace]};
        #erl_mfa{m=M,f=F,a=A} ->
            FileName = get_filename(M),
            %% {Line,[{{M,F},{A,Args},[{file,FileName}]} | Trace]};
            %% {Line,[{{M,F},[A | Args],[{file,FileName}]} | Trace]};
            {Line,[{{M,F,A},Args,[{file,FileName}]} | Trace]};
        Other ->
            {Line,[{Other,Args,[{file,<<"-no-file-">>},{line,Line}]} | Trace]}
    end;
do_stackframe(#current_line{line=Line}, {_,Trace}, _St) ->
    {Line,Trace};
do_stackframe(#loop_frame{}, Acc, _St) ->       %Ignore these
    Acc.

get_filename(Mod) ->
    Comp = erlang:get_module_info(Mod, compile),
    case lists:keyfind(source, 1, Comp) of
        {source,FileName} ->
            BaseName = filename:basename(FileName),
            list_to_binary(BaseName);
        false ->                                %The compiler doesn't know
            <<"-no-file-">>
    end.

%% Define IS_MAP/1 macro for is_map/1 bif.
-ifdef(HAS_MAPS).
-define(IS_MAP(T), is_map(T)).
-else.
-define(IS_MAP(T), false).
-endif.

%% encode_list([Term], State) -> {[LuerlTerm],State}.
%% encode(Term, State) -> {LuerlTerm,State}.

encode_list(Ts, St) ->
    lists:mapfoldl(fun encode/2, St, Ts).

encode(nil, St) -> {nil,St};
encode(false, St) -> {false,St};
encode(true, St) -> {true,St};
encode(B, St) when is_binary(B) -> {B,St};
encode(A, St) when is_atom(A) -> {atom_to_binary(A, utf8),St};
encode(N, St) when is_number(N) -> {N,St};      %Integers and floats
encode(F, St) when ?IS_MAP(F) -> encode(maps:to_list(F), St);
encode(L, St0) when is_list(L) ->
    %% Encode the table elements in the list.
    EncTab = fun ({K0,V0}, {I,S0}) ->
                     {K1,S1} = encode(K0, S0),
                     {V1,S2} = encode(V0, S1),
                     {{K1,V1},{I,S2}};
                 (V0, {I,S0}) ->
                     {V1,S1} = encode(V0, S0),
                     {{I,V1},{I+1,S1}}
             end,
    {Es,{_,St1}} = lists:mapfoldl(EncTab, {1,St0}, L),
    {T,St2} = luerl_heap:alloc_table(Es, St1),
    {T,St2};                                    %No more to do for now
encode(F, St) when is_function(F, 2) ->
    F1 = fun(Args, State) -> F(Args, State) end,
    %% io:format("enc ~p\n", [#erl_func{code=F1}]),
    {#erl_func{code=F1}, St};
encode(F, St) when is_function(F, 1) ->
    F1 = fun(Args, State) -> Res = F(Args), {Res,State} end,
    %% io:format("enc ~p\n", [#erl_func{code=F1}]),
    {#erl_func{code=F1}, St};
encode({M,F,A}, St) when is_atom(M) and is_atom(F) ->
    %% io:format("enc ~p\n", [#erl_mfa{m=M,f=F,a=A}]),
    {#erl_mfa{m=M,f=F,a=A}, St};
encode({userdata,Data}, St) ->
    luerl_heap:alloc_userdata(Data, St);
%% % Table refs should not be re-encoded
%% encode(#tref{}=T, St) ->
%%     case luerl_heap:chk_table(T, St) of
%%         ok -> {T, St};
%%         error -> error(badarg)
%%     end;
encode(Term, _) -> error({badarg,Term}).        %Can't encode anything else

%% decode_list([LuerlTerm], State) -> [Term].
%% decode(LuerlTerm, State) -> Term.
%%  In decode we track of which tables we have seen to detect
%%  recursive references and generate an error when that occurs.

decode_list(Lts, St) ->
    lists:map(fun (Lt) -> decode(Lt, St) end, Lts).

decode(LT, St) ->
    decode(LT, St, []).

decode(nil, _, _) -> nil;
decode(false, _, _) -> false;
decode(true, _, _) -> true;
decode(B, _, _) when is_binary(B) -> B;
decode(A, _, _) when is_atom(A) -> atom_to_binary(A, utf8);
decode(N, _, _) when is_number(N) -> N;         %Integers and floats
decode(#tref{}=T, St, In) ->
    decode_table(T, St, In);
decode(#usdref{}=U, St, In) ->
    decode_userdata(U, St, In);
decode(#funref{}=Fun, St, In) ->
    decode_luafunc(Fun, St, In);
decode(#erl_func{}=Fun, St, In) ->
    decode_erlfunc(Fun, St, In);
decode(#erl_mfa{}=Mfa, St, In) ->
    decode_erlmfa(Mfa, St, In);
decode(Lua, _, _) -> error({badarg,Lua}).       %Shouldn't have anything else

decode_table(#tref{i=N}=T, St, In0) ->
    case lists:member(N, In0) of
        true -> error({recursive_table,T});     %Been here before
        false ->
            In1 = [N|In0],                      %We are in this as well
            case luerl_heap:get_table(T, St) of
                #table{a=Arr,d=Dict} ->
                    Fun = fun (K, V, Acc) ->
                                  [{decode(K, St, In1),decode(V, St, In1)}|Acc]
                          end,
                    Ts = ttdict:fold(Fun, [], Dict),
                    array:sparse_foldr(Fun, Ts, Arr);
                _Undefined -> error(badarg)
            end
    end.

decode_userdata(U, St, _In) ->
    {#userdata{d=Data},_} = luerl_heap:get_userdata(U, St),
    {userdata,Data}.

decode_luafunc(Fun, _St, _In) ->
    %% io:format("dec ~p\n", [Fun]),
    fun(Args, State) ->
            luerl_emul:functioncall(Fun, Args, State)
    end.

decode_erlfunc(#erl_func{code=Fun}=_Ef, _St, _In) ->
    %% io:format("dec ~p\n", [Ef]),
    Fun.                                        %Just the bare fun

decode_erlmfa(#erl_mfa{m=Mod,f=Func,a=Arg}=_Mfa, _St, _In) ->
    %% io:format("mfa ~p\n", [Mfa]),
    {Mod,Func,Arg}.

%% Externalize and Internalize ensure that the VM state passed in
%% can be stored externally or can be recreated from external storage.
%% Currently very simple: only random state needs special treatment.

externalize(S) ->
    luerl_lib_math:externalize(S).

internalize(S) ->
    luerl_lib_math:internalize(S).

%% put_private(Key, Value, State) ->
%%   State.
%% get_private(Key, State) ->
%%   Value.
%% delete_private(Key, State) ->
%%   Value.
put_private(Key, Value, S) ->
    Private = maps:put(Key, Value, S#luerl.private),
    S#luerl{private=Private}.

get_private(Key, S) ->
    maps:get(Key, S#luerl.private).

delete_private(Key, S) ->
    Private = maps:remove(Key, S#luerl.private),
    S#luerl{private=Private}.
